src/Servers/Kestrel/Core/src/Internal/Http3/Http3Connection.cs

```diff
@@ -176,38 +176,64 @@ public void Tick(DateTimeOffset now)
+            // 1. When a stream first starts and waits to receive headers.
+            //    Uses RequestHeadersTimeout.
+            // 2. When a stream finished and is waiting for underlying transport to drain.
+            //    Uses MinResponseDataRate.
+
             var ticks = now.Ticks;
 
             lock (_streams)
             {
                 foreach (var stream in _streams.Values)
                 {
-                    if (stream.ReceivedHeader)
+                    if (stream.IsReceivingHeader)
                     {
-                        continue;
-                    }
+                        if (stream.StreamTimeoutTicks == default)
+                        {
+                            // On expiration overflow, use max value.
+                            var expirationTicks = ticks + _context.ServiceContext.ServerOptions.Limits.RequestHeadersTimeout.Ticks;
+                            stream.StreamTimeoutTicks = expirationTicks >= 0 ? expirationTicks : long.MaxValue;
+                        }
 
-                    if (stream.HeaderTimeoutTicks == default)
-                    {
-                        // On expiration overflow, use max value.
-                        var expirationTicks = ticks + _context.ServiceContext.ServerOptions.Limits.RequestHeadersTimeout.Ticks;
-                        stream.HeaderTimeoutTicks = expirationTicks >= 0 ? expirationTicks : long.MaxValue;
+                        if (stream.StreamTimeoutTicks < ticks)
+                        {
+                            if (stream.IsRequestStream)
+                            {
+                                stream.Abort(new ConnectionAbortedException(CoreStrings.BadRequest_RequestHeadersTimeout), Http3ErrorCode.RequestRejected);
+                            }
+                            else
+                            {
+                                stream.Abort(new ConnectionAbortedException(CoreStrings.Http3ControlStreamHeaderTimeout), Http3ErrorCode.StreamCreationError);
+                            }
+                        }
                     }
-
-                    if (stream.HeaderTimeoutTicks < ticks)
+                    else if (stream.IsDraining)
                     {
-                        if (stream.IsRequestStream)
+                        var minDataRate = _context.ServiceContext.ServerOptions.Limits.MinResponseDataRate;
+                        if (minDataRate == null)
                         {
-                            stream.Abort(new ConnectionAbortedException(CoreStrings.BadRequest_RequestHeadersTimeout), Http3ErrorCode.RequestRejected);
+                            continue;
                         }
-                        else
+
+                        if (stream.StreamTimeoutTicks == default)
+                        {
+                            var expirationTicks = Math.Max(
+                                _context.TimeoutControl.GetWriteTimingTimeoutTimestamp(),
```

Comment:
I think I prefer calculating the drain timeout this way vs deriving the worst-case from `MaxResponseBufferSize` like we do for HTTP/1.x and HTTP/2. Can we do the same thing in `Http1Connection` and `Http2Connection` for consistency?

Reply:
TBH I thought grabbing the internally calculated timestamp like this was a bit of a hack. I was expecting feedback from you to tell me how it should really work 😋

Reply:
My knee-jerk reaction was to have you copy the existing logic, but this makes some sense. Maybe we could move this `Math.Max` logic into `TimeoutControl` and expose a `GetResponseDrainDeadline()` method instead of `GetWriteTimingTimeoutTimestamp()` so it feels less hacky.